### PR TITLE
[gcs] Fix actor killing race condition

### LINKED ISSUE #17369: [Core] Actor hangs after other actors are killed
This very simple script is hanging for me. This is on nightly wheels.

```
import ray
ray.init(num_cpus=2)

class BaseWorker:
    """A class to execute arbitrary functions. Does not hold any state."""

    def execute(self, func):
        """Executes the input function and returns the output.
        Args:
            func(Callable): A function that does not take any arguments.
        """
        return func()

remote_cls = ray.remote(num_cpus=1)(BaseWorker)
workers = [remote_cls.remote() for _ in range(2)]

for worker in workers:
	ray.kill(worker)

remote_cls = ray.remote(num_cpus=1)(BaseWorker)
workers = [remote_cls.remote() for _ in range(2)]

func  = lambda: 1
ray.get([w.execute.remote(func) for w in workers]) # This hangs!
```

### PR DESCRIPTION
**Pull request by fishbone**
## Why are these changes needed?
Creating and killing an actor immediately will have a race condition.

1. Lease 
2. Canceling
3. Lease reply

It will leave the worker leak. This PR fixed it.

## Checks
Closes #17369 
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tests/test_basic_3.py
Comment (rkooo567) [inline python/ray/tests/test_basic_3.py:198]: Can you actually test with `ray.kill(worker_1, no_restart=False)` and create an actor with `max_restarts=-1`, and see if this works as expected?

**Review by rkooo567 — COMMENTED** [review]
LGTM, but I wonder if this works properly when the max_restarts > 1. Can you verify this scenario as well?

File: python/ray/tests/test_basic_3.py
Comment (iycheng) [reply python/ray/tests/test_basic_3.py:198]: With no_restart=False it'll hang. Without, it'll exit as expected.

Comment (iycheng) [reply python/ray/tests/test_basic_3.py:198]: what do you think is the correct behavior here for the setup?

Comment (rkooo567) [reply python/ray/tests/test_basic_3.py:198]: Hmm I think actor should be restarted until restart happens `max_restarts ` times. So you can test with 

`no_restart=False` & max_restarts=1

and make sure it is alive the first time, and killed in the second time (then that's the most correct behavior)

Comment (iycheng) [reply python/ray/tests/test_basic_3.py:198]: It'll be alive the first time, but still not being killed the second time :( 
Let me investigate more.

**Comment by rkooo567** [comment]
Hey @ffbin can you review this PR as well?

File: src/ray/gcs/gcs_server/gcs_actor_scheduler.cc
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_scheduler.cc:443]: Why don't we make it as a private method?

File: src/ray/gcs/gcs_server/gcs_actor_scheduler.cc
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_scheduler.cc:466]: Can you write a comment what's the condition of this, and why we call kill worker?

File: src/ray/gcs/gcs_server/gcs_actor_scheduler.cc
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_scheduler.cc:452]: This RPC will fail most of time right? Wonder if we can have DEBUG log here for the future debugging (in case this introduces any bug)

File: src/ray/gcs/gcs_server/gcs_actor_scheduler.cc
Comment (rkooo567) [inline src/ray/gcs/gcs_server/gcs_actor_scheduler.cc:496]: Can you also explain when this happens in the comment?

**Review by rkooo567 — APPROVED** [review]
LGTM. A couple minor comments.

**Comment by rkooo567** [comment]
Btw, before we merge it, can you run the stress test; `test_many_dead_actors`?

File: src/ray/gcs/gcs_server/gcs_actor_scheduler.cc
Comment (iycheng) [reply src/ray/gcs/gcs_server/gcs_actor_scheduler.cc:443]: I think this kill_worker should be scoped to this function only. The reason is that it's actually coupling with this one: it set the actor id = Nil. It means the worker is leased, but not assigned any actor.

In actor_manager, there actually is a function to kill an actor.

File: src/ray/gcs/gcs_server/gcs_actor_scheduler.cc
Comment (iycheng) [reply src/ray/gcs/gcs_server/gcs_actor_scheduler.cc:452]: What do you mean here by "fail"? Is it because it's killing the actor? I think it depends on the implementation of client-side code. Ideally, it shouldn't.

Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_actor_scheduler.cc:452]: when I say fail, I meant the RPC returns the error status because the actor is already killed. So, my hypothesis here is that most of time actors are already killed, and in some situations, they are not killed, and this code will ensure the correct "final state". Did I understand this part correctly? I'd like to add logs in this case. 

```cpp
 RAY_UNUSED(cli->KillActor(request, [](Status &status) { RAY_LOG(DEBUG) << "Killing an actor because the lease request is returned after the actor is killed. Killing status: " << status}));
```

File: src/ray/gcs/gcs_server/gcs_actor_scheduler.cc
Comment (rkooo567) [reply src/ray/gcs/gcs_server/gcs_actor_scheduler.cc:443]: That makes sense. Can you comment this here (so we can avoid confusion in the future)?

**Review by rkooo567 — APPROVED** [review]

**Comment by rkooo567** [comment]
Let me ping @ffbin in person.

**Comment by ffbin** [comment]
> Hey @ffbin can you review this PR as well?

Ok, i will take a look today.

File: src/ray/gcs/gcs_server/gcs_actor_scheduler.cc
Comment (ffbin) [inline src/ray/gcs/gcs_server/gcs_actor_scheduler.cc:454]: Add a blank line to make the code look cleaner?

File: src/ray/gcs/gcs_server/gcs_actor_scheduler.cc
Comment (ffbin) [inline src/ray/gcs/gcs_server/gcs_actor_scheduler.cc:465]: If actor is dead, why we still need to kill actor?

File: src/ray/gcs/gcs_server/gcs_actor_scheduler.cc
Comment (iycheng) [reply src/ray/gcs/gcs_server/gcs_actor_scheduler.cc:454]: Sorry I missed this comment. :(

File: src/ray/gcs/gcs_server/gcs_actor_scheduler.cc
Comment (iycheng) [reply src/ray/gcs/gcs_server/gcs_actor_scheduler.cc:465]: The reason is that there is a race condition:

sending leasing request  -> successfully leased, response sending back (not received) -> kill actor (actor dead) -> receive the response[the place you comment].

**Comment by rkooo567** [comment]
There's a couple follow up comment that is not addressed yet. @iycheng will create a small follow up PR

**Pull request merged**
